CmdStanMCMC/MLE/VB from read_cmdstan_csv ouput

jgabry opened the pull request:
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and agree to license (see below)

#### Summary

Closes #411 

Allows creating CmdStanMCMC, CmdStanVB, and CmdStanMLE objects from the list returned by `read_cmdstan_csv()`. 
In reality these objects are not exactly the same because they don't have an associated CmdStanRun object. This means that methods like `cmdstan_diagnose()` and `save_output_files()` are unavailable, but the most important methods (e.g., `draws()`, `summary()`) work fine. 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 
**Columbia University**


By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

jgabry commented on R/fit.R:
```diff
@@ -15,9 +15,6 @@ CmdStanFit <- R6::R6Class(
       self$runset$num_procs()
     },
     print = function(variables = NULL, ..., digits = 2, max_rows = 10) {
-      if (!length(self$output_files(include_failed = FALSE))) {
-        stop("Fitting failed. Unable to print.", call. = FALSE)
-      }
```
Need to remove this so it doesn't error if output files aren't available. Otherwise we can't use this with the output from read_cmdstan_csv(). 

However, this doesn't risk `print()` failing to catch an important error because the same error will be caught by the `metadata()` method that is called internally by `print()`. `metadata()` is smarter about throwing the error because it only throws the error if the files are missing AND `metadata()` hasn't already been read in.

jgabry commented on R/fit.R:
```diff
@@ -330,7 +327,7 @@ CmdStanFit$set("public", name = "lp", value = lp)
 summary <- function(variables = NULL, ...) {
   draws <- self$draws(variables)
-  if (self$runset$method() == "sample") {
+  if (self$metadata()$method == "sample") {
```
We need to check the method via metadata()$method and not runset$method() or it won't work with the objects created after read_cmdstan_csv

jgabry commented on R/fit.R:
```diff
@@ -599,10 +596,10 @@ CmdStanFit$set("public", name = "output", value = output)
-  if (!length(self$output_files(include_failed = FALSE))) {
-    stop("Fitting failed. Unable to retrieve the metadata.", call. = FALSE)
-  }
   if (is.null(private$metadata_)) {
+    if (!length(self$output_files(include_failed = FALSE))) {
+      stop("Fitting failed. Unable to retrieve the metadata.", call. = FALSE)
+    }
```
Move this error after the `is.null` check so that it's only thrown if metadata hasn't already been read in.

jgabry commented on R/fit.R:
```diff
@@ -705,17 +702,17 @@ CmdStanMCMC <- R6::R6Class(
         if (self$runset$args$validate_csv) {
           fixed_param <- runset$args$method_args$fixed_param
-          data_csv <- read_cmdstan_csv(
+          csv_contents <- read_cmdstan_csv(
```
This and every other change to `csv_contents` isn't important, it just makes naming consistent throughout our code

jgabry commented on R/fit.R:
```diff
@@ -730,9 +727,6 @@ CmdStanMCMC <- R6::R6Class(
 
     # override the CmdStanFit draws method
     draws = function(variables = NULL, inc_warmup = FALSE) {
-      if (!length(self$output_files(include_failed = FALSE))) {
```
This error will also be thrown by `private$read_csv_` so it is unnecessary here. We also need to remove it or it will result in errors after creating the objects from read_cmdstan_csv.

jgabry commented on R/fit.R:
```diff
@@ -773,45 +767,45 @@ CmdStanMCMC <- R6::R6Class(
         stop("No chains finished successfully. Unable to retrieve the draws.", call. = FALSE)
       }
-      data_csv <- read_cmdstan_csv(
+      csv_contents <- read_cmdstan_csv(
```
this all and cases below are just name changes

jgabry commented on R/read_csv.R:
```diff
@@ -370,9 +373,112 @@ read_sample_csv <- function(files,
+#' x <- read_cmdstan_csv(fit$output_files())
+#' fit <- as_cmdstan_mcmc(x)
+#' }
+as_cmdstan_mcmc <- function(csv_contents, check_diagnostics = TRUE) {
```
What do you think of the names `as_cmdstan_mcmc()`, `as_cmdstan_mle()`, `as_cmdstan_vb()`?

rok-cesnovar replied:
I like them, but have another idea. What about `as_cmdstan_fit()` that does all three?

jgabry replied:
Good idea. I just changed it.

jgabry commented on R/read_csv.R:
```diff
@@ -370,9 +373,112 @@ read_sample_csv <- function(files,
+#'
+#' @noRd
+#'
+CmdStanMCMC2 <- R6::R6Class(
+  classname = "CmdStanMCMC2",
+  inherit = CmdStanMCMC,
```
Don't actually create CmdStanMCMC objects but CmdStanMCMC2 objects that inherit from CmdStanMCMC. Same approach with MLE and VB below. This avoids having to create the runset object because we can override the `initialize` method.  

Suggestions for better class name than `CmdStanMCMC2`?

rok-cesnovar replied:
What aobut CmdStanMCMCCSV, CmdStanMCMC_CSV? I am bad for names.

jgabry commented on R/read_csv.R:
```diff
@@ -370,9 +373,112 @@ read_sample_csv <- function(files,
+  )
+)
+
+error_unavailable_CmdStanFit2 <- function(...) {
+  fun <- switch(class(self)[1],
+                CmdStanMCMC2 = "as_cmdstan_mcmc()",
+                CmdStanMLE2 = "as_cmdstan_mle()",
+                CmdStanVB2 = "as_cmdstan_vb()")
+  stop("This method is not available for objects created using ", fun, ".",
+       call. = FALSE)
+}
+unavailable_methods_CmdStanFit2 <-
+  c(
+    paste0("cmdstan_", c("diagnose", "summary")),
+    paste0(c("output", "latent_dynamics"), "_files"),
+    paste("save_", c("output", "latent_dynamics"), "_files"),
+    "data_file",
+    "save_data_file",
+    "init",
+    "output",
+    "return_codes",
+    "time"
+  )
+for (method in unavailable_methods_CmdStanFit2) {
+  CmdStanMCMC2$set("public", name = method, value = error_unavailable_CmdStanFit2)
+  CmdStanMLE2$set("public", name = method, value = error_unavailable_CmdStanFit2)
+  CmdStanVB2$set("public", name = method, value = error_unavailable_CmdStanFit2)
+}
```
This makes sure that the methods that won't work have informative error messages.

jgabry reviewed (commented):
@rok-cesnovar Just added a bunch of review comments that hopefully help explain the changes.

jgabry commented on DESCRIPTION:
```diff
@@ -1,6 +1,6 @@
 Package: cmdstanr
 Title: R Interface to 'CmdStan'
-Version: 0.3.0
+Version: 0.3.0.9000
```
bumping to dev version

jgabry commented on R/read_csv.R:
```diff
@@ -370,9 +388,124 @@ read_sample_csv <- function(files,
+#'
+#' @noRd
+#'
+CmdStanMCMC2 <- R6::R6Class(
```
I put these in the read_csv.R file because of the use case but we could also but these in the fit.R file with the definition of CmdStanMCMC, CmdStanMLE, etc. Preference?

rok-cesnovar replied:
I think having them here is fine.

jgabry commented:
@yizhang-yiz I think this is now ready to play around with if you want to try it out.

yizhang-yiz commented:
```r
> fit <- cmdstanr::as_cmdstan_mcmc(dir(pattern=".csv", full.name=TRUE))
```
This works
```r
bayesplot::mcmc_dens_overlay(fit$draws(), pars = c("CL", "Q", "V1", "V2", "ka", "sigma"), facet_args = list(nrow = 2))
```
but this doesn't
```r
> print(fit$time(), digits = 3)
Error: This method is not available for objects created using as_cmdstan_mcmc().
```
I wonder what the rationale is.

jgabry commented:
> I wonder what the rationale is.

No good one other than it would take more effort to get that to work. `read_cmdstan_csv()` doesn't currently give you the timing information. When model fitting happens we create a CmdStanRun object

https://github.com/stan-dev/cmdstanr/blob/46ae80f12881356e9152edd3881457b6da40016f/R/model.R#L804-L806

which that handles preparing the call to cmdstan, setting up the external processes, and then running cmdstan. It's also the object that report the run times (we don't wait for reading in the csv to report the time). The way I created `as_cmdstan_mcmc()` it avoids creating this object entirely and just uses the output from `read_cmdstan_csv()`. So to get time information with `as_cmdstan_mcmc()` we'd just need to add it to `read_cmdstan_csv()` (probably a good idea anyway).

rok-cesnovar commented:
> we'd just need to add it to read_cmdstan_csv() (probably a good idea anyway). 

Agree. Will do that separately.

jgabry commented:
I changed to `as_cmdstan_fit()`, good idea. I haven't changed the class names yet but I'm leaning towards `CmdStanMCMC_CSV`. That's definitely better than `CmdStanMCMC2`. 

Let's wait until  #414 is merged and then we can update this PR to include `time` before merging it.

jgabry commented:
Ok I changed the class names to `CmdStanMCMC_CSV`, `CmdStanMLE_CSV`, `CmdStanVB_CSV` and I added the `time()` method for `CmdStanMCMC_CSV` after pulling in the changes from #414.

rok-cesnovar reviewed (approved):
Looks good to me. Thanks.
